[Peh Hoe Khim Marcus] iP

marcuspeh opened the pull request:
# DUKE
> "My job is not to be easy on people. My job is to make them better." [Source](https://www.treasurequotes.com/quotes/my-job-is-not-to-be-easy-on-people-my-job-is-2)

Duke helps you remember things and keep tracks of things for you.😉

### Main _advantage_ of Duke includes
- Text based
- Easy to use
- Minimalistic

### To get Duke working,
1. Download the [here](https://github.com/marcuspeh/ip/releases/download/A-Jar/A-Jar.jar)
2. To run the file, type in `java -jar A-Jar.jar` in the command line.
3. Add your tasks.
4. Let it manage the task for u.

### Features:
- [X] Managing Task
- [X] Managing datelines
- [X] Managing Todos
- [ ] Reminders

### Miscellaneous
If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Duke.run();
    }
}
```

Commit:
Add Gradle support

marcuspeh committed:
testing git using intelij

marcuspeh committed:
Add greeting, echo and exit

marcuspeh committed:
update methods to camelCase notation

marcuspeh committed:
Add the ability to store whatever text entered by the user and display them back to the user

marcuspeh committed:
Change stream to IntStream in order to remove the for loop and speed up the codes

marcuspeh committed:
Update class variables for task to private

marcuspeh committed:
Add mark as done

marcuspeh committed:
Fix minor code stlying problems

marcuspeh committed:
Add support for 3 task - ToDos, Events and Deadlines

marcuspeh committed:
Update message when list is empty

marcuspeh committed:
Update testing files

marcuspeh committed:
Catch specific exception instead of generic ones

marcuspeh committed:
Add support for deleting tasks from the list

marcuspeh committed:
Update testing codes for delete

marcuspeh committed:
Update the version for java docs

marcuspeh committed:
Replace static class variables with enum

marcuspeh committed:
Implement save and load task functionality

marcuspeh committed:
Update Duke to recognise date and time

marcuspeh committed:
Update test-ui files

marcuspeh committed:
Update output of dates in task

marcuspeh committed:
Update test-ui to include dateTime errors

marcuspeh committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadlines.java
#	src/main/java/Duke.java
#	src/main/java/Events.java
#	src/main/java/Task.java

marcuspeh committed:
Extract parser, storage and task list class

marcuspeh committed:
Extract Ui to deal with interactions with user

marcuspeh committed:
Update javadocs

marcuspeh committed:
Extract command out as classes

marcuspeh committed:
Organize into packages

marcuspeh committed:
Update javadocs statements

marcuspeh committed:
Add test files for package duke.task

marcuspeh committed:
Add test for parser and duke exception

marcuspeh committed:
Update javadocs and headers

marcuspeh committed:
Add support to find task

marcuspeh committed:
Move keyword enum to util package

marcuspeh committed:
Update javadocs

marcuspeh committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/util/Keyword.java
#	src/main/java/duke/util/TaskList.java

marcuspeh committed:
Update README

marcuspeh committed:
Update gitignore

marcuspeh committed:
Merge branch 'add-gradle-support'

marcuspeh committed:
Update README

marcuspeh committed:
Add checkstyle to project

marcuspeh committed:
Modify codes to checkstyle suggestions

marcuspeh committed:
Add support for javaFx

marcuspeh committed:
Abstract message as a class of its own

marcuspeh committed:
Create the GUI for duke with fxml and javafx

marcuspeh committed:
Add in welcome message for GUI

marcuspeh committed:
Update GUI UI

koh-jx commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,43 @@
+     * Adds a new task to the task list.
+     *
+     * @param taskList duke.main.TaskList to execute the command.
+     * @param ui       To interact with the user.
```
Looks like there may be unintentional whitespaces here?

marcuspeh replied:
I think we are supposed to leave whitespace to match the description with the next line. I remember reading it somewhere, although i cant find it anymore. Will have a look at the conventions again and changed if needed.

koh-jx replied:
Oh, now that you mention it, it does makes sense. Ignore my suggestion then! 😆

koh-jx commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,36 @@
+     * Finds all the task containing the keywords.
+     *
+     * @param taskList duke.main.TaskList to execute the command.
+     * @param ui       To interact with the user.
```
Similar to the unintentional whitespace earlier. Minor documentation layout error

koh-jx commented on src/main/java/duke/command/TodoCommand.java:
```diff
@@ -0,0 +1,42 @@
+    /** Stores the message entered by the user. */
+    private String message;
+
+    /** Constructor for duke.task.Deadlines.TodoCommand.
```
Opening /** should be on a separate line

marcuspeh replied:
Yup, I agree that for the opening of /**, it should be on a new line. However, if it is for a class method, it is an exception and can be on the same line.
![image](https://user-images.githubusercontent.com/54053213/131599690-fad2739e-7a61-4955-bd12-f2bd12997089.png)

koh-jx replied:
Ahh I see, I missed this out, thanks for pointing this out

koh-jx commented on src/main/java/duke/util/DukeException.java:
```diff
@@ -0,0 +1,14 @@
+ * @version A-MoreOOP
+ * @since 21 Aug 2021
+ */
+public class DukeException extends Exception {
```
What this Exception is used for is not very clear at first glance. Perhaps more explanation on how it may be used may help.

marcuspeh replied:
That's a good suggestion. I will write more explanation for this.

koh-jx commented on src/main/java/duke/util/Keyword.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    Keyword(String keyword) {
+        this.keyword = keyword;
+        this.separator = "";
+        this.saveWord = "";
```
You may opt to "Avoid unnecessary use of this with fields"
![image](https://user-images.githubusercontent.com/39460640/131525953-54b9de7d-06fd-47e7-86d2-7d4998b87148.png)

marcuspeh replied:
Thanks for alerting me on this. Totally missed this out.

koh-jx commented on src/main/java/duke/util/Keyword.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    Keyword(String keyword, String separator, String saveWord) {
+        this.keyword = keyword;
+        this.separator = separator;
+        this.saveWord = saveWord;
```
Same as above, may opt to avoid unnecessary use of this with fields if you wish

marcuspeh replied:
Got it! Thanks for spotting it in my codes.

koh-jx commented on src/main/java/duke/util/Message.java:
```diff
@@ -0,0 +1,29 @@
+
+    /**
+     * Constructor for Message.
+     * @param s Message returned by the command.
```
Empty line between description and parameter section

marcuspeh replied:
Ops, I missed it out. Will have it changed.

marcuspeh committed:
Fix codes base on the suggestion of checkstyle

koh-jx commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,106 @@
+            file.getParentFile().mkdirs();
+            file.createNewFile();
+
+            FileWriter fileWrite = new FileWriter(file);
+            BufferedWriter bufferedWriter = new BufferedWriter(fileWrite);
```
Really small issue, but I'm curious as to why BufferedWriter was initialised as bufferedWriter, but FileWriter fileWrite didn't follow this convention as well (fileWrite vs fileWriter).

Then again, since fileWrite isn't used anywhere else, you may also possibly consider instantiating it directly within bufferedWriter? Same with fileReader in the same class, but this is probably entirely up to preference.

marcuspeh replied:
Now that you pointed it out, I agree with what you mentioned about instantiating it directly. As for the naming convention, you raised a good point too.

koh-jx commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @return message to be used by either the graphic UI or command line UI.
+     */
+    public Message chatErrorMessage() {
```
Not a coding standard issue, but you may want to abstract out the list of commands since both chatErrorMessage( ) and importTaskErrorMessage( ) uses it. If not a developer may end up updating one without updating the other.

marcuspeh replied:
Thanks for the suggestion, I will have it updated bases on your suggestion.

koh-jx reviewed (commented):
I mostly had only feedback on pretty minor details. Really concise and very clean! 😃

marcuspeh committed:
Update files base on comments given by PR

erwinqxy commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,93 @@
+import duke.util.TaskList;
+import duke.util.Ui;
+
+public class Duke {
```
Hi! Maybe you can write a descriptive header comment for Duke here

marcuspeh replied:
Thanks for your suggestion. I will add it in in the next push

erwinqxy commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,93 @@
+        ui.exitMessage().printMessage();
+    }
+
+    public String getResponse(String message) {
```
Hi! Maybe you can write a descriptive header comment for  `getResponse` here

marcuspeh replied:
Thanks for the suggestion. I will add it in.

erwinqxy commented on src/main/java/duke/util/Keyword.java:
```diff
@@ -0,0 +1,79 @@
+ * @version Level-9
+ * @since 19 Aug 2021
+ */
+public enum Keyword {
```
good use of enums here! _(Code Quality - Make the Code Obvious)_

marcuspeh replied:
Thank you :)

erwinqxy commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,68 @@
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+/**
```
Not sure if you should have a line break here?

marcuspeh replied:
Ops, I missed that out. Will have it changed

erwinqxy commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,44 @@
+    @Override
+    public Message execute(TaskList taskList, Ui ui) {
+        try {
+            return taskList.deleteTask(Integer.parseInt(message.substring(Keyword.DELETE.length() + 1)));
```
Hi! Do you want to consider adding line breaks here to improve readability?

marcuspeh replied:
Hmm where do u recommend adding in the line break?

erwinqxy commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,36 @@
+import duke.util.TaskList;
+import duke.util.Ui;
+
+public class FindCommand implements Command {
```
hi! i think you can add a descriptive header for your `FIndCommand `class here

marcuspeh replied:
Got it! Will have it updated

erwinqxy commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,36 @@
+    @Override
+    public Message execute(TaskList taskList, Ui ui) {
+        List<Task> allTask = taskList.getTaskList();
+        String[] task = IntStream.range(0, allTask.size())
```
Good practice in using line breaks! Easy to read! :smile:

marcuspeh replied:
Thank you :)

erwinqxy commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,44 @@
+     * @return the type of command as a enum duke.util.Keyword.
+     * @throws DukeException if the command is not recognised.
+     */
+    public static Command parseChat(String message) throws DukeException {
```
Good that you avoided arrowhead style code! 👍🏻

erwinqxy commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,44 @@
+        }
+        throw new DukeException("Command not parsable");
+    }
+}
```
The structure here is clean and very readable!

erwinqxy commented on src/main/java/duke/task/Deadlines.java:
```diff
@@ -0,0 +1,51 @@
+
+    /**
+     * Saves the task to the txt file.
+     * Format is as follow: {@literal <}Type(D){@literal >} | {@literal <}Description{@literal >} |
```
I appreciate this documentation. Thank you!

erwinqxy replied:
Explain WHAT, not HOW _(Code Quality)_

marcuspeh replied:
Thats great! Thank you :)

erwinqxy commented on src/main/java/duke/task/ToDos.java:
```diff
@@ -0,0 +1,45 @@
+     * @return string to save the txt file.
+     */
+    public String saveOutput() {
+        return String.format("T | %s | %s", super.getTask(), super.getIsDone() ? 1 : 0);
```
I think you can add some line breaks here to improve readability

marcuspeh replied:
Thats a good suggestion. Even though the character count is less than 110, i think a line break is necessary here

erwinqxy commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+     */
+    public Message markDone(int n) {
+        Task task = taskList.get(n - 1);
+        boolean success = task.markDone();
```
i think `success` can be renamed to something more clearer to sound like a boolean. perhaps `isSuccess`? What do you think?

marcuspeh replied:
That's a good suggestion. Thanks!

erwinqxy reviewed (commented):
Overall, I like the structure and flow of your code. And I think you have adhered to most of the Java coding standards and the Code Quality. Good job! 👍🏻

marcuspeh committed:
Update codes base on PR review

marcuspeh committed:
Update styling of dialogue box

marcuspeh committed:
Add in assertion to check for assumptions about the program state

The checks is done during runtime.

Checking for bugs during runtime is important during production. A failure indicates a possible bug in the code because the code has resulted in a program state that violates an assumption about how the code should behave.

Let's add in assertion to test for bugs during runtime.

Assertion is introduced into the necceary parts of the codes to check for those assumptions.

marcuspeh committed:
Update codes to conform with coding standards

Ui clas contains methods that start without a verb.

The changing of the method names will conform with the coding standard.

Let's add in a ```format``` before each of the method name for message.

marcuspeh committed:
Update main application class name for gradle

The main application class name for build.grade is Duke.

Duke runs the command line user interface while Launcher runs the GUI version. Since javaFx is implemented, the command line version should not be the default one to run.

Let's replace the application class name in build.gradle with Launcher.

Launcher is preferable than Duke as it provides a GUI for the app.

marcuspeh committed:
Update comments for ListCommand

marcuspeh committed:
Merge pull request #3 from marcuspeh/branch-A-Streams

Update comments for ListCommand

marcuspeh committed:
Merge branch 'master' into branch-A-CodeQuality

marcuspeh committed:
Merge pull request #2 from marcuspeh/branch-A-CodeQuality

Update codes to conform with coding standards

marcuspeh committed:
Merge branch 'master' into branch-A-Assertions

marcuspeh committed:
Merge pull request #1 from marcuspeh/branch-A-Assertions

Add in assertion to check for assumptions about the program state

marcuspeh committed:
Add github actions files

marcuspeh committed:
Create manual.yml

marcuspeh committed:
Merge branch 'master' of https://github.com/marcuspeh/ip

marcuspeh committed:
Update CI files

marcuspeh committed:
Revert "Update CI files"

This reverts commit 4d8f5e9de23d741db3db7fa45ea6fa728534c9f5.

marcuspeh committed:
Delete manual.yml

marcuspeh committed:
Update gradle.yml

marcuspeh committed:
Update gradle.yml

marcuspeh committed:
Add in help command

marcuspeh committed:
Add in help command

marcuspeh committed:
Merge branch 'branch-BCD-Extension'

marcuspeh committed:
Polish Gui

marcuspeh committed:
Update manifest

marcuspeh committed:
Update user guide

marcuspeh committed:
Update user docs

marcuspeh committed:
Update comments for header

marcuspeh committed:
Update testing methods

marcuspeh committed:
Update exit command to work

marcuspeh committed:
Update userguide

marcuspeh committed:
Update README.md
